### DISCUSSION TITLE
Get static mesh from CSG nodes at runtime

### DISCUSSION THREAD

**Discussion by sammonius** [godotengine/godot-proposals#7440] [3D]
Just a function like `csg_node.construct_static_mesh()`. It would fix the performance issues with CSG in most cases.

Or this might be pretty cool:

- The addition operator on 2 meshes performs a union
- The subtraction operator gets the difference or subtraction
- The "&" operator gets the intersersection (assuming "&" is an implementable operator)

Just wanted to discuss how this could be done. I think it's a super useful feature, like for adding bullet holes or digging.

**Comment by theraot** [comment]
Does calling [`get_meshes`](https://docs.godotengine.org/en/stable/classes/class_csgshape3d.html#class-csgshape3d-method-get-meshes) on your `CSGCombiner3D` not work for you? You can create new `MeshInstance3D` or `CSGMesh3D` and assign them the meshes you get from it.

**Reply by sammonius** [reply]
Oh thanks, I wasn't aware of that function.

**Comment by Calinou** [comment]
- This is already being tracked in https://github.com/godotengine/godot-proposals/issues/182.

> It would fix the performance issues with CSG in most cases.

This will not improve CSG construction or rendering performance. In fact, converting to mesh after every change will likely make it even slower if you're looking to modify geometry during gameplay (which is often a bad idea with CSG, for more reasons than just performance).

The main use case for converting CSG to mesh is to use it in a MeshInstance3D, GridMap or GPUParticles3D, if you need features exclusive to those nodes.
